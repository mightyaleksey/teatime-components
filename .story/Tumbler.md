## Preview


### size s

```
<div>
  some text for the baseline
  <Tumbler
    defaultChecked={true}
    name='tumbler-s'
    size='s'/>
  {' '}
  <Tumbler
    defaultChecked={false}
    disabled={true}
    name='tumbler-s-d'
    size='s'/>
</div>
```


### size m

```
<div>
  some text for the baseline
  <Tumbler
    name='tumbler-m'
    size='m'/>
  {' '}
  <Tumbler
    disabled={true}
    name='tumbler-m-d'
    size='m'/>
</div>
```


### size l

```
<div>
  some text for the baseline
  <Tumbler
    name='tumbler-l'
    offText='Down'
    onText='Up'
    size='l'/>
  {' '}
  <Tumbler
    defaultChecked={true}
    disabled={true}
    name='tumbler-l-d'
    offText='Down'
    onText='Up'
    size='l'/>
</div>
```
