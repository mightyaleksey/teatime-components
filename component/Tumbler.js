'use strict';

const {Component, PropTypes} = require('react');
const {genericName, omitNonStandardAttrs} = require('../lib/util');
const {genericId} = require('../lib/identity');
const {isUndefined, noop} = require('../lib/dash');
const React = require('react');
const cc = require('classnames');

const cssModules = {
  l: require('../style/tumbler/tumbler-l.css'),
  m: require('../style/tumbler/tumbler-m.css'),
  s: require('../style/tumbler/tumbler-s.css'),
};

class Tumbler extends Component {
  constructor(props) {
    super(props);

    this.state = {
      id: props.id || genericId(),
    };
  }

  componentWillReceiveProps(nextProps) {
    this.setState({
      id: isUndefined(nextProps.id)
        ? this.state.id
        : nextProps.id,
    });
  }

  css = tokenName => genericName(this.props, tokenName)

  focus() {
    if (!this._input) return;
    this._input.focus();
  }

  _onChange = e => {
    const {checked, value} = e.target;

    this.props.onChange(e, {checked, value});
  }

  render() {
    const {className, offText, onText, ...props} = this.props;
    const {id} = this.state;

    const {css} = this;

    // why this looks like a pie,
    // see: http://stackoverflow.com/questions/9273016/why-is-this-inline-block-element-pushed-downward
    return (
      <span
        {...omitNonStandardAttrs(props)}
        className={cc(css('wrapper'), className)}
        onChange={void 0}>
        <span className={css('baseline')}>
          <div className={css('shape')}>
            {this.renderInput()}
            <label
              className={css('control')}
              htmlFor={id}>
              <span className={css('label')}>{onText}</span>
              <span className={css('label')}>{offText}</span>
              <span className={css('delimiter')}>&nbsp;</span>
            </label>
          </div>
        </span>
      </span>
    );
  }

  renderInput() {
    const {
      checked,
      defaultChecked,
      disabled,
      onBlur,
      onFocus,
      name,
      value,
    } = this.props;

    const {id} = this.state;

    const {css} = this;

    return (
      <input
        checked={checked}
        className={css('native')}
        defaultChecked={defaultChecked}
        disabled={disabled}
        id={id}
        name={name}
        onBlur={onBlur}
        onChange={this._onChange}
        onFocus={onFocus}
        ref={r => this._input = r}
        type='checkbox'
        value={value}/>
    );
  }
}

Tumbler.defaultProps = {
  offText: 'Off',
  onText: 'On',
  onChange: noop,
  size: 'm',
  styles: cssModules,
};

Tumbler.propTypes = {
  checked: PropTypes.bool,
  className: PropTypes.string,
  defaultChecked: PropTypes.bool,
  disabled: PropTypes.bool,
  id: PropTypes.string,
  name: PropTypes.string.isRequired,
  offText: PropTypes.string,
  onBlur: React.PropTypes.func,
  onChange: React.PropTypes.func,
  onFocus: React.PropTypes.func,
  onText: PropTypes.string,
  size: PropTypes.oneOf([
    'l',
    'm',
    's',
  ]),
  styles: PropTypes.object,
  value: PropTypes.string,
};

module.exports = Tumbler;
