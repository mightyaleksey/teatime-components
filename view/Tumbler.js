'use strict';

const { Component, PropTypes } = require('react');
const { bind, sanitizeProps } = require('../tool/component');
const { generateId } = require('../tool/identity');
const { noop } = require('../tool/func');
const { styleName } = require('../tool/className');
const React = require('react');

class Tumbler extends Component {
  constructor(props) {
    super(props);

    bind(this, 'onChange');

    this.state = {
      id: this.props.id || generateId(),
    };
  }

  componentWillReceiveProps({ id }) {
    if (id) {
      this.setState({id});
    }
  }

  onChange(e) {
    const { checked, value } = e.target;
    this.props.onChange(e, {checked, value});
  }

  render() {
    const { off, on, styles, ...o } = this.props;
    const { id } = this.state;

    /**
     * Still there is an issue about controlled and uncontrolled components,
     * related to the input[type="checkbox"] and input[type="radio"].
     * It results in the way controlled components are determined.
     *
     * @see https://github.com/facebook/react/blob/v15.1.0/src/renderers/dom/client/wrappers/ReactDOMInput.js#L171
     * @see https://github.com/facebook/react/issues/6779
     */

    return (
      <div className={styleName(this.props)}>
        <input
          type='checkbox'
          {...sanitizeProps(['styleName'], o)}
          className={styles.native}
          id={id}
          onChange={this.onChange}/>
        <label className={styles.control} htmlFor={id}>
          <span className={styles.label}>{on}</span>
          <span className={styles.label}>{off}</span>
          <span className={styles.delimiter}>&nbsp;</span>
        </label>
      </div>
    );
  }
}

Tumbler.defaultProps = {
  off: 'Off',
  on: 'On',
  onChange: noop,
  styleName: 'wrapper',
  styles: {},
};

Tumbler.propTypes = {
  name: PropTypes.string.isRequired,
  off: PropTypes.string,
  on: PropTypes.string,
  onChange: PropTypes.func,
  styleName: PropTypes.string,
  styles: PropTypes.shape({
    control: PropTypes.string.isRequired,
    delimiter: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    native: PropTypes.string.isRequired,
    wrapper: PropTypes.string,
  }),
};

module.exports = Tumbler;
