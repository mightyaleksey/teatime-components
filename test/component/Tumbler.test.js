import React from 'react';
import Tumbler from '../../component/Tumbler';
import renderer from 'react-test-renderer';

test('Tumbler renders correctly with defaults', () => {
  const tree = renderer.create(
    <Tumbler
      name='tumbler control'/>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

test('Tumbler renders correctly with defined props', () => {
  const tree = renderer.create(
    <Tumbler
      checked={true}
      disabled={true}
      id='tumblers id'
      name='tumbler control'
      offText='~~OFF~~'
      onText='~~ON~~'/>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
