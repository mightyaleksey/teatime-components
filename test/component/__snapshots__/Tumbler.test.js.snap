exports[`test Tumbler renders correctly with defaults 1`] = `
<span
  className="wrapper"
  onChange={undefined}>
  <span
    className="baseline">
    <div
      className="shape">
      <input
        checked={undefined}
        className="native"
        defaultChecked={undefined}
        disabled={undefined}
        id="_teatime1"
        name="tumbler control"
        onChange={[Function]}
        type="checkbox"
        value={undefined} />
      <label
        className="control"
        htmlFor="_teatime1">
        <span
          className="label">
          On
        </span>
        <span
          className="label">
          Off
        </span>
        <span
          className="delimiter">
           
        </span>
      </label>
    </div>
  </span>
</span>
`;

exports[`test Tumbler renders correctly with defined props 1`] = `
<span
  className="wrapper"
  onChange={undefined}>
  <span
    className="baseline">
    <div
      className="shape">
      <input
        checked={true}
        className="native"
        defaultChecked={undefined}
        disabled={true}
        id="tumblers id"
        name="tumbler control"
        onChange={[Function]}
        type="checkbox"
        value={undefined} />
      <label
        className="control"
        htmlFor="tumblers id">
        <span
          className="label">
          ~~ON~~
        </span>
        <span
          className="label">
          ~~OFF~~
        </span>
        <span
          className="delimiter">
           
        </span>
      </label>
    </div>
  </span>
</span>
`;
